File: ai_diagnostic/sensitivity.py
"""
Sensitivity analysis for the Bayesian diagnostic framework.

Provides four analyses:
  1. Parameter sensitivity — how much each evidence parameter contributes
  2. Flip analysis — which missing evidence, if found, would change the diagnosis
  3. Tipping point — minimum evidence value at which the diagnosis flips
  4. Prior sensitivity — effect of perturbing priors ±10%
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional

from .evidence import Evidence, EVIDENCE_PARAMS
from .bayesian import BayesianDiagnostic, DiagnosisResult
from .likelihoods import LikelihoodTable, FAILURE_CLASSES


@dataclass
class SensitivityReport:
    """Results of a full sensitivity analysis."""

    parameter_sensitivity: Dict[str, float] = field(default_factory=dict)
    flip_analysis: List[Dict[str, object]] = field(default_factory=list)
    tipping_points: Dict[str, Optional[float]] = field(default_factory=dict)
    prior_sensitivity: Dict[str, float] = field(default_factory=dict)

    def summary(self) -> str:
        lines = [
            "=" * 60,
            "  SENSITIVITY ANALYSIS",
            "=" * 60,
        ]

        lines.append("")
        lines.append("-" * 60)
        lines.append("  Parameter Sensitivity (|delta in top posterior|)")
        lines.append("-" * 60)
        if self.parameter_sensitivity:
            sorted_params = sorted(
                self.parameter_sensitivity.items(), key=lambda x: abs(x[1]), reverse=True
            )
            for name, delta in sorted_params:
                lines.append(f"  {name:<50s} {delta:>+8.4f}")
        else:
            lines.append("  (none)")

        lines.append("")
        lines.append("-" * 60)
        lines.append("  Flip Analysis (missing evidence that would change diagnosis)")
        lines.append("-" * 60)
        if self.flip_analysis:
            for entry in self.flip_analysis:
                lines.append(
                    f"  {entry['parameter']:<40s} "
                    f"→ {entry['alternative_diagnosis']:<25s} "
                    f"conf={entry['new_confidence']:.2%}"
                )
        else:
            lines.append("  (no missing evidence that would flip diagnosis)")

        lines.append("")
        lines.append("-" * 60)
        lines.append("  Tipping Points (evidence value at which diagnosis flips)")
        lines.append("-" * 60)
        if self.tipping_points:
            for name, val in self.tipping_points.items():
                if val is None:
                    lines.append(f"  {name:<50s}   (no flip in [0, 1])")
                else:
                    lines.append(f"  {name:<50s}   {val:.4f}")
        else:
            lines.append("  (none)")

        lines.append("")
        lines.append("-" * 60)
        lines.append("  Prior Sensitivity (|delta| for ±10% perturbation)")
        lines.append("-" * 60)
        if self.prior_sensitivity:
            sorted_priors = sorted(
                self.prior_sensitivity.items(), key=lambda x: abs(x[1]), reverse=True
            )
            for cause, delta in sorted_priors:
                lines.append(f"  {cause:<50s} {delta:>+8.4f}")
        else:
            lines.append("  (none)")

        lines.append("=" * 60)
        return "\n".join(lines)


def _rediagnose(
    diagnostic: BayesianDiagnostic,
    evidence: Evidence,
    prior_overrides: Optional[Dict[str, float]] = None,
) -> DiagnosisResult:
    return diagnostic.diagnose(evidence, prior_overrides=prior_overrides)


def _compute_parameter_sensitivity(
    result: DiagnosisResult,
    diagnostic: BayesianDiagnostic,
) -> Dict[str, float]:
    """For each evidence param, set to 0 and measure change in top posterior."""
    top_diagnosis = result.diagnosis
    base_confidence = result.confidence
    evidence = result.evidence_vector
    sensitivity: Dict[str, float] = {}

    for param in EVIDENCE_PARAMS:
        current_val = getattr(evidence, param)
        if current_val == 0.0:
            sensitivity[param] = 0.0
            continue
        zeroed = Evidence(**{**evidence.to_dict(), param: 0.0})
        new_result = _rediagnose(diagnostic, zeroed)
        sensitivity[param] = new_result.posteriors.get(top_diagnosis, 0.0) - base_confidence

    return sensitivity


def _compute_flip_analysis(
    result: DiagnosisResult,
    diagnostic: BayesianDiagnostic,
) -> List[Dict[str, object]]:
    """For each evidence param at 0, set to 1 and check if diagnosis changes."""
    evidence = result.evidence_vector
    current_diagnosis = result.diagnosis
    flips: List[Dict[str, object]] = []

    for param in EVIDENCE_PARAMS:
        current_val = getattr(evidence, param)
        if current_val != 0.0:
            continue
        one_val = Evidence(**{**evidence.to_dict(), param: 1.0})
        new_result = _rediagnose(diagnostic, one_val)
        if new_result.diagnosis != current_diagnosis:
            flips.append({
                "parameter": param,
                "current_value": current_val,
                "alternative_diagnosis": new_result.diagnosis,
                "new_confidence": new_result.confidence,
            })

    return flips


def _compute_tipping_points(
    result: DiagnosisResult,
    diagnostic: BayesianDiagnostic,
) -> Dict[str, Optional[float]]:
    """For each evidence param, find minimum value at which diagnosis flips."""
    top_diagnosis = result.diagnosis
    evidence = result.evidence_vector
    tipping: Dict[str, Optional[float]] = {}

    for param in EVIDENCE_PARAMS:
        current_val = getattr(evidence, param)
        if current_val >= 1.0:
            tipping[param] = None
            continue

        lo, hi = current_val, 1.0
        found = None

        for _ in range(50):
            mid = (lo + hi) / 2.0
            probe = Evidence(**{**evidence.to_dict(), param: mid})
            probe_result = _rediagnose(diagnostic, probe)
            if probe_result.diagnosis != top_diagnosis:
                found = mid
                hi = mid
            else:
                lo = mid
            if hi - lo < 1e-6:
                break

        tipping[param] = found

    return tipping


def _compute_prior_sensitivity(
    result: DiagnosisResult,
    diagnostic: BayesianDiagnostic,
) -> Dict[str, float]:
    """Perturb each prior ±10% and measure change in top posterior."""
    top_diagnosis = result.diagnosis
    base_confidence = result.confidence
    evidence = result.evidence_vector
    sensitivity: Dict[str, float] = {}

    for cause in FAILURE_CLASSES:
        base_prior = diagnostic.table.prior(cause)
        perturbed_prior = min(base_prior * 1.1, 1.0)
        overrides = {cause: perturbed_prior}
        up_result = _rediagnose(diagnostic, evidence, prior_overrides=overrides)
        delta = up_result.posteriors.get(top_diagnosis, 0.0) - base_confidence
        sensitivity[cause] = delta

    return sensitivity


def compute_sensitivity(
    result: DiagnosisResult,
    diagnostic: BayesianDiagnostic,
) -> SensitivityReport:
    """
    Compute all four sensitivity analyses for a given diagnosis.

    Args:
        result: A completed DiagnosisResult.
        diagnostic: The BayesianDiagnostic engine used to produce it.

    Returns:
        A SensitivityReport with parameter sensitivity, flip analysis,
        tipping points, and prior sensitivity.
    """
    return SensitivityReport(
        parameter_sensitivity=_compute_parameter_sensitivity(result, diagnostic),
        flip_analysis=_compute_flip_analysis(result, diagnostic),
        tipping_points=_compute_tipping_points(result, diagnostic),
        prior_sensitivity=_compute_prior_sensitivity(result, diagnostic),
    )

File: ai_diagnostic_rs/src/sensitivity.rs
//! Sensitivity analysis: parameter impact, flip analysis, tipping points,
//! prior perturbation.

use crate::bayesian::{BayesianDiagnostic, DiagnosisResult};
use crate::evidence::EVIDENCE_PARAMS;
use crate::likelihoods::FAILURE_CLASSES;
use serde_json::{json, Map, Value};

#[derive(Debug, Clone, serde::Serialize)]
pub struct FlipEntry {
    pub parameter: String,
    pub current_value: f64,
    pub alternative_diagnosis: String,
    pub new_confidence: f64,
}

#[derive(Debug, Clone, Default, serde::Serialize)]
pub struct SensitivityReport {
    pub parameter_sensitivity: Vec<(String, f64)>,
    pub flip_analysis: Vec<FlipEntry>,
    pub tipping_points: Vec<(String, Option<f64>)>,
    pub prior_sensitivity: Vec<(String, f64)>,
}

impl SensitivityReport {
    fn sorted<'a, I>(items: I) -> Vec<I::Item>
    where
        I: IntoIterator<Item = (&'a str, f64)>,
    {
        let mut v: Vec<_> = items.into_iter().collect();
        v.sort_by(|a, b| b.1.abs().partial_cmp(&a.1.abs()).unwrap_or(std::cmp::Ordering::Equal));
        v
    }

    pub fn summary(&self) -> String {
        let mut lines = vec![
            "=".repeat(60),
            "  SENSITIVITY ANALYSIS".to_string(),
            "=".repeat(60),
            String::new(),
            "-".repeat(60),
            "  Parameter Sensitivity (|delta in top posterior|)".to_string(),
            "-".repeat(60),
        ];
        if self.parameter_sensitivity.is_empty() {
            lines.push("  (none)".to_string());
        } else {
            for (name, delta) in Self::sorted(
                self.parameter_sensitivity
                    .iter()
                    .map(|(n, d)| (n.as_str(), *d)),
            ) {
                lines.push(format!("  {name:<50} {delta:>+8.4}"));
            }
        }

        lines.push(String::new());
        lines.push("-".repeat(60));
        lines.push(
            "  Flip Analysis (missing evidence that would change diagnosis)".to_string(),
        );
        lines.push("-".repeat(60));
        if self.flip_analysis.is_empty() {
            lines.push("  (no missing evidence that would flip diagnosis)".to_string());
        } else {
            for e in &self.flip_analysis {
                lines.push(format!(
                    "  {:<40} → {:<25} conf={:.2}",
                    e.parameter,
                    e.alternative_diagnosis,
                    e.new_confidence * 100.0
                ));
            }
        }

        lines.push(String::new());
        lines.push("-".repeat(60));
        lines.push("  Tipping Points (evidence value at which diagnosis flips)".to_string());
        lines.push("-".repeat(60));
        if self.tipping_points.is_empty() {
            lines.push("  (none)".to_string());
        } else {
            for (name, val) in &self.tipping_points {
                match val {
                    None => lines.push(format!("  {name:<50}   (no flip in [0, 1])")),
                    Some(v) => lines.push(format!("  {name:<50}   {v:.4}")),
                }
            }
        }

        lines.push(String::new());
        lines.push("-".repeat(60));
        lines.push("  Prior Sensitivity (|delta| for ±10% perturbation)".to_string());
        lines.push("-".repeat(60));
        if self.prior_sensitivity.is_empty() {
            lines.push("  (none)".to_string());
        } else {
            for (cause, delta) in Self::sorted(
                self.prior_sensitivity.iter().map(|(c, d)| (c.as_str(), *d)),
            ) {
                lines.push(format!("  {cause:<50} {delta:>+8.4}"));
            }
        }

        lines.push("=".repeat(60));
        lines.join("\n")
    }

    pub fn to_dict(&self) -> Map<String, Value> {
        let mut m = Map::new();
        m.insert(
            "parameter_sensitivity".into(),
            Value::Object(
                self.parameter_sensitivity
                    .iter()
                    .map(|(k, v)| (k.clone(), json!(v)))
                    .collect(),
            ),
        );
        m.insert(
            "flip_analysis".into(),
            json!(self.flip_analysis),
        );
        m.insert(
            "tipping_points".into(),
            Value::Object(
                self.tipping_points
                    .iter()
                    .map(|(k, v)| (k.clone(), json!(v)))
                    .collect(),
            ),
        );
        m.insert(
            "prior_sensitivity".into(),
            Value::Object(
                self.prior_sensitivity
                    .iter()
                    .map(|(k, v)| (k.clone(), json!(v)))
                    .collect(),
            ),
        );
        m
    }
}

fn top_of(result: &DiagnosisResult) -> f64 {
    result.posteriors
        .iter()
        .find(|(c, _)| *c == result.diagnosis)
        .map(|(_, v)| *v)
        .unwrap_or(0.0)
}

/// Parameter sensitivity: zero each present param, measure delta in top posterior.
fn compute_parameter_sensitivity(
    result: &DiagnosisResult,
    diagnostic: &BayesianDiagnostic,
) -> Result<Vec<(String, f64)>, String> {
    let base_confidence = result.confidence;
    let evidence = &result.evidence_vector;
    let mut out = Vec::new();

    for param in EVIDENCE_PARAMS {
        if evidence.get(param).unwrap() == 0.0 {
            out.push((param.to_string(), 0.0));
            continue;
        }
        let mut ev = evidence.clone();
        ev.set(param, 0.0)?;
        let new_result = diagnostic.diagnose(&ev, None, 0.6)?;
        out.push((param.to_string(), top_of(&new_result) - base_confidence));
    }
    Ok(out)
}

/// Flip analysis: set each absent param to 1.0; record diagnosis changes.
fn compute_flip_analysis(
    result: &DiagnosisResult,
    diagnostic: &BayesianDiagnostic,
) -> Result<Vec<FlipEntry>, String> {
    let evidence = &result.evidence_vector;
    let current_diagnosis = &result.diagnosis;
    let mut flips = Vec::new();

    for param in EVIDENCE_PARAMS {
        if evidence.get(param).unwrap() != 0.0 {
            continue;
        }
        let mut ev = evidence.clone();
        ev.set(param, 1.0)?;
        let new_result = diagnostic.diagnose(&ev, None, 0.6)?;
        if &new_result.diagnosis != current_diagnosis {
            flips.push(FlipEntry {
                parameter: param.to_string(),
                current_value: 0.0,
                alternative_diagnosis: new_result.diagnosis.clone(),
                new_confidence: new_result.confidence,
            });
        }
    }
    Ok(flips)
}

/// Tipping points: bisect each param toward 1.0 to find flip threshold.
fn compute_tipping_points(
    result: &DiagnosisResult,
    diagnostic: &BayesianDiagnostic,
) -> Result<Vec<(String, Option<f64>)>, String> {
    let evidence = &result.evidence_vector;
    let top_diagnosis = &result.diagnosis;
    let mut tipping = Vec::new();

    for param in EVIDENCE_PARAMS {
        let current = evidence.get(param).unwrap();
        if current >= 1.0 {
            tipping.push((param.to_string(), None));
            continue;
        }
        let (mut lo, mut hi) = (current, 1.0f64);
        let mut found: Option<f64> = None;

        for _ in 0..50 {
            let mid = (lo + hi) / 2.0;
            let mut ev = evidence.clone();
            ev.set(param, mid)?;
            let probe = diagnostic.diagnose(&ev, None, 0.6)?;
            if probe.diagnosis != *top_diagnosis {
                found = Some(mid);
                hi = mid;
            } else {
                lo = mid;
            }
            if hi - lo < 1e-6 {
                break;
            }
        }
        tipping.push((param.to_string(), found));
    }
    Ok(tipping)
}

/// Prior sensitivity: perturb each prior +10% (capped at 1.0), measure delta.
fn compute_prior_sensitivity(
    result: &DiagnosisResult,
    diagnostic: &BayesianDiagnostic,
) -> Result<Vec<(String, f64)>, String> {
    let base_confidence = result.confidence;
    let evidence = &result.evidence_vector;
    let mut out = Vec::new();

    for cause in FAILURE_CLASSES {
        let base_prior = diagnostic.table.prior(cause).unwrap_or(0.0);
        let overrides = std::collections::HashMap::from([(
            cause.to_string(),
            (base_prior * 1.1).min(1.0),
        )]);
        let up_result = diagnostic.diagnose(evidence, Some(&overrides), 0.6)?;
        let delta = top_of(&up_result)
            - base_confidence;
        out.push((cause.to_string(), delta));
    }
    Ok(out)
}

/// Compute all four sensitivity analyses for a given diagnosis.
///
/// Sub-diagnoses use the engine's default table, mirroring the reference
/// implementation's behavior.
pub fn compute_sensitivity(
    result: &DiagnosisResult,
    diagnostic: &BayesianDiagnostic,
) -> Result<SensitivityReport, String> {
    Ok(SensitivityReport {
        parameter_sensitivity: compute_parameter_sensitivity(result, diagnostic)?,
        flip_analysis: compute_flip_analysis(result, diagnostic)?,
        tipping_points: compute_tipping_points(result, diagnostic)?,
        prior_sensitivity: compute_prior_sensitivity(result, diagnostic)?,
    })
}
